Add parse function as a new entry point

robinkrahl opened the pull request:
Previously, there were three different entry points for html2text:
from_read, from_read_with_decorator and from_read_rich.  This had two
drawbacks:

Firstly, the user always had to parse and render the HTML at the same
time.  But there are use cases where the user wants to render the same
HTML document using different widths or decorators.  This involved
parsing the document multiple times.

Secondly, it was impossible to create rich text with a custom decorator.
This would have required a new function from_read_rich_with_decorator.

This patch introduces a new entry point for the pubic API, the parse
function.  It reads and parses the HTML document and returns a
RenderTree object.  This object can then be rendered into a RenderedText
object using a custom decorator or the default plain and rich
decorators.  The RendererText object can be converted to a string or to
a vector of annotated lines.

----

Since I brought up the rich text decorator issue (#24), I noticed another problem:  When displaying an HTML document in a terminal user interface using cursive, I have to render the document multiple times during the layout phase and if the terminal size changes.  Currently, I have to parse the document every time I want to render it although it did not change.  This is not very efficient, especially as the HTML parsing seems to be the most expensive step.  This PR attemps to solve both issues.

jugglerchris commented:
This all looks good and more flexible in principle, thanks!  That would be useful for my project (a terminal mail client) too.

However the tests are all failing - looks like `from_read()` is returning an empty string.  Are you able to look into that please?

Also, I guess it would be good to test the new functions.

robinkrahl commented:
I fixed the test failures.  Just before I pushed the changes, I removed a clone that seemed redundant without re-running the tests.  Apparently it was *not* redundant.  I also had to remove the `From` implementations for `RenderedText` conversions to enable compilation with Rust versions earlier than 1.41.0.

Regarding the tests for the new functions, most of them are already covered by the existing tests.  My suggestion would be to add new tests for these cases:
- Rendering the same `RenderTree` multiple times with different decorators and widths.
- Rendering rich text with a custom decorator.

Does that make sense?

jugglerchris commented:
Wow, I agree that the clone looks redundant.  I don't quite understand why it isn't.

Those tests you propose sound perfect - I agree that the functions are otherwise tested already indirectly.

Thanks for the improvements!

robinkrahl commented:
I added some basic test cases.  I had to derive `Clone` for the render tree to be able to render the same tree multiple times, and `PartialEq` for the tagged types to make it possible to write unit tests for them.

jugglerchris commented:
Great, thanks!
The `Clone` will be because I think it destructively walks the render tree (as there wasn't any reason not to).  Maybe that could be revisited if the render tree might now live longer.

jugglerchris commented:
Is it useful to you to publish this to crates.io soon?

robinkrahl commented:
Thanks for merging!  No, this is not urgent for me.  While html2text can be quite slow in debug mode, I didn’t have any performance issues in release mode.  It just felt wrong to parse the same document every time I update the layout. ;-)
